### PR TITLE
Stripe checkout customization cookbook

### PR DESCRIPTION
**Pull request by bruno-ds**
This new cookbook explain how to:
- override layout & templates
- Extend Payum's gateway in order to add metadata

| Q               | A
| --------------- | -----
| Branch?         | 1.2
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | fixes https://github.com/Sylius/Sylius/pull/9640
| License         | MIT

Hello, this PR is the sequel of https://github.com/Sylius/Sylius/pull/9640 (this one was too complicated to merge, I restarted it from 0).
It build successfully with sphinx & I took into account each of your comments except for creating an 1.3 version: I do not use 1.3 and would not be able to test it for the moment.

Those changes are in qualification in our integration servers and will be deployed in the next release of https://store.itophub.io

### CODE REVIEW

File: docs/cookbook/payments/stripe-customization.rst [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+        /**
+         * @var Context $context
+         */
+        public function onExecute(Context $context)
+        {
+            if (!$this->supports($context)) {
+                return;
+            }
+
+            /** @var $request Capture */
```
Comment (pamil) [inline docs/cookbook/payments/stripe-customization.rst]:
```suggestion
            /** @var Capture $request */
```

File: docs/cookbook/payments/stripe-customization.rst [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+.. code-block:: yaml
+
+    # app/config/config.yml
+    twig:
+        paths:
+            "%kernel.root_dir%/../app/Resources/PayumCore/views": PayumCore
+            "%kernel.root_dir%/../app/Resources/PayumStripe/views": PayumStripe
+
+Having this configuration defined, you can now create your customuzations in your `app/Ressource/` directory, as usual.
```
Comment (pamil) [inline docs/cookbook/payments/stripe-customization.rst]:
```suggestion
Having this configuration defined, you can now create your customizations in your `app/Resources/` directory, as usual.
```

File: docs/cookbook/payments/stripe-customization.rst [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+                    data-key="{{ publishable_key }}"
+                    data-image="{{ asset('path/to/image.png') }}"
+                    data-name="{{ model.name|default("Your store name") }}"
+                    data-description="{{ model.description|default("") }}"
+                    data-amount="{{ model.amount }}"
+                    data-currency="{{ model.currency|default("EUR") }}"
+                    data-email="{{ app.user.email }}"
+                    data-locale="auto"
+                    data-allow-remember-me="false"
+
```
Comment (pamil) [inline docs/cookbook/payments/stripe-customization.rst]:
```suggestion
```

File: docs/cookbook/payments/stripe-customization.rst [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+    /**
+     *
+     * This class handles the addition of metadata to Stripe
+     *
+     * It is wired using Payum's built in extension system.
+     * It is wired to Stripe_checkout only using a factory filter on the service tag.
+     * It is additionally filtered to the Capture request using a code check.
+     *
+     * @author Bruno DA SILVA (https://github.com/bruno-ds)
```
Comment (pamil) [inline docs/cookbook/payments/stripe-customization.rst]: We usually don't mention authors in codeblocks in our docs.

File: docs/cookbook/payments/stripe-customization.rst [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+     */
+    class StripeAddMetadataOnCaptureExtensions implements ExtensionInterface
+    {
+        /**
+         * @var Context $context
```
Comment (pamil) [inline docs/cookbook/payments/stripe-customization.rst]: Could you remove these kinds of docblocks (as there are typehints already)?

**Comment by CoderMaggie** [comment]
Hi Bruno! 👋 I like the approach of slimming the giant PRs, thanks! Take care of the redundant lines and some proofreading for the misspellings. In the meantime we will try to check the usability of this cookbook (by simply trying to follow it on our own). 

Nevertheless, you can consider what @pamil has suggested to you. 💪

File: docs/cookbook/payments/stripe-customization.rst (hunk 5 above)
Comment (Zales0123) [reply docs/cookbook/payments/stripe-customization.rst]: Typehints in function body are 👍 IMO, especially in the documentation where you cannot jump to the code and check the return value. This one is redundant due to typehints in function parameters. I would only change them to interfaces, whenever it's possible (e.g. `Order` -> `OrderInterface`).

**Comment by bruno-ds** [comment]
@pamil :  
 - thank you for your feedbacks, I've taken them into account
 - I don't know if you want me to get ride of the typehints in the functions body (cf https://github.com/Sylius/Sylius/pull/9880#discussion_r233839314)
 - For the plugin question : 
    - This code still rely on Payum's integration (it do not leverage a standalone capability)
    - The needs will probably vary greatly and this is more a proposal than a final implementation (personally I added several custom fields)

@CoderMaggie : thank you, I'll wait for your feedback :)

**Comment by bruno-ds** [comment]
friendly ping to @CoderMaggie :)

**Comment by bruno-ds** [comment]
Friendly (but last) ping to @CoderMaggie  &  @pamil :)

**Pull request closed without merging**